pkg/beaconfqdn/dissector.go

```diff
@@ -0,0 +1,174 @@
+		dissectChannel:    make(chan *hostname.FqdnInput),
+	}
+}
+
+//collect sends a chunk of data to be analyzed
+func (d *dissector) collect(datum *hostname.FqdnInput) {
+	d.dissectChannel <- datum
+}
+
+//close waits for the collector to finish
+func (d *dissector) close() {
+	close(d.dissectChannel)
+	d.dissectWg.Wait()
+	// d.closedCallback()
+}
+
+//start kicks off a new analysis thread
+func (d *dissector) start() {
+	d.dissectWg.Add(1)
+	go func() {
+		ssn := d.db.Session.Copy()
+		defer ssn.Close()
+
+		for datum := range d.dissectChannel {
+
+			// set up src match key (src ip + network uuid)
+			srcMatchKey := datum.Src.SrcBSONKey()
+
+			var dstList []interface{}
+			// create dst match query section
+			for _, dst := range datum.ResolvedIPs {
+				dstList = append(dstList, dst.DstBSONKey())
+			}
+
+			// This will work for both updating and inserting completely new Beacons
+			// for every new hostnames record we have, we will check every entry in the
+			// uconn table where the source IP from the hostnames record connected to one
+			// of the associated IPs for  FQDN. This
+			// will always return a result because even with a brand new database, we already
+			// created the uconns table. It will only continue and analyze if the connection
+			// meets the required specs, again working for both an update and a new src-fqdn
+			// pair. We would have to perform this check regardless if we want the rolling
+			// update option to remain, and this gets us the vetting for both situations, and
+			// Only works on the current entries - not a re-aggregation on the whole collection,
+			// and individual lookups like this are really fast. This also ensures a unique
+			// set of timestamps for analysis.
+			uconnFindQuery := []bson.M{
+				bson.M{"$match": srcMatchKey},
+				bson.M{"$match": bson.M{"$or": dstList}},
+				bson.M{"$project": bson.M{
+					"src": 1,
+					"ts": bson.M{
+						"$reduce": bson.M{
+							"input":        "$dat.ts",
+							"initialValue": []interface{}{},
+							"in":           bson.M{"$concatArrays": []interface{}{"$$value", "$$this"}},
+						},
+					},
+					"bytes": bson.M{
+						"$reduce": bson.M{
+							"input":        "$dat.bytes",
+							"initialValue": []interface{}{},
+							"in":           bson.M{"$concatArrays": []interface{}{"$$value", "$$this"}},
+						},
+					},
+					"count":  bson.M{"$sum": "$dat.count"},
+					"tbytes": bson.M{"$sum": "$dat.tbytes"},
+					"icerts": bson.M{"$anyElementTrue": []interface{}{"$dat.icerts"}},
+				}},
+				bson.M{"$group": bson.M{
+					"_id":    "$src",
+					"ts":     bson.M{"$push": "$ts"},
+					"bytes":  bson.M{"$push": "$bytes"},
+					"count":  bson.M{"$sum": "$count"},
+					"tbytes": bson.M{"$sum": "$tbytes"},
+					"icerts": bson.M{"$push": "$icerts"},
+				}},
+				bson.M{"$unwind": "$ts"},
+				bson.M{"$unwind": "$ts"},
+				bson.M{"$group": bson.M{
+					"_id":    "$_id",
+					"ts":     bson.M{"$addToSet": "$ts"},
```

Comment:
We are, the timestamps need to be unique for the rest of the calculation. This is the same as in the regular beaconing. There are frequently many connections that end up having the same timestamp and that will skew the results since the distance between them will be zero.

Reply:
Sounds good. We might consider adding comments to the beaconing/ beaconFQDN packages to explain this. :+1: